### ISSUE TITLE
Attached arrows and lines (glue points)

### ISSUE THREAD

**Issue by hmaurer** [excalidraw/excalidraw#157]
Hello there, nice work so far!

I was wondering if the ability to "attach" the start and end of arrows to shapes is something that would be useful for this little app, or whether it should be considered out of scope (could potentially open a can of worms). If so, in addition to "free arrows" (arrows that are not tied to any shapes, as you have now), you would also have arrows where one (or both) ends are tied to a shape, and move along with it.

**Comment by vjeux** [comment]
I think that would be valuable. Otherwise you have to move everything in a repetitive manner when you want to make edits.

**Comment by hmaurer** [comment]
@vjeux right, and that gets tiring very quickly :) One thing that could add a bit of complexity (though there might be a simple solution) is making sure it works nicely with whatever approach is taken for #148 .

**Comment by vjeux** [comment]
One thing I don’t like with most implementations is that you only have 3 control points (left, middle, right) and if you have more than a few connectors, then they all are clamped together. I’d love to spread them out all over the surface we’re trying to connect to. Should be easy to implement and greatly improve the rendering.

Another pet peeve of mine is that the arrow is directly attached to the element. I usually give some padding.

**Comment by vjeux** [comment]
I don't think we need to wait for having curved arrows to fix this, if something wants to give it a try, would be a welcome addition :)

**Comment by gaearon** [comment]
Let's just make sure it's not overly strict about attachment. For example, I often move the arrow a little bit if it allows to avoid intersecting it with another bubble that just happens to be on the way. But strict attachment snap systems make this a pain. Keynote is particularly annoying in that sense.

**Comment by hmaurer** [comment]
@gaearon I can relate to that; it has been a major pet peeve of mine with some of those diagram editors. What would be a good (yet simple) rule to handle attachments while leaving some freedom to the user with placement?

A simple rule with padding could be "draw a straight line between the centre of the two shapes, and place the endpoints of the arrow on the intersections between that straight line and slightly bigger clones of said two shapes. Here's a diagram because it's hard to put in plain English (made on excalidraw, very meta):

![excalidraw](https://user-images.githubusercontent.com/541045/71784652-88607980-2fed-11ea-9f36-a91401c65183.png)

The grey shapes are meant to show how padding is added; they wouldn't actually be on the diagram.

Now if you want to leave freedom to the user on the placement of the endpoints, how do you go about this? You could let them drag the endpoint anywhere on the grey outlines, but then what do you do if the shape is moved? Leave the endpoint in the same spot?

**Comment by hmaurer** [comment]
Follow up idea: we implement the basic rule I described above, but instead of "straight line between the two shape centres" we generalise to "straight line between shape centre and next control point". If we implement a basic version of curved arrows as @vjeux suggested, which wouldn't actually curve the arrows, just break them up into multiple segments intersecting at angles, then this would likely give enough freedom to the user to work around the annoyances you described @gaearon . A diagram to explain:

![excalidraw (1)](https://user-images.githubusercontent.com/541045/71784978-06268400-2ff2-11ea-9b05-6a78444117d9.png)

(the first segment should be a line; I used an arrow because excalidraw doesn't have lines...)

So basically this would keep endpoint placement fully automatic.

**Comment by hmaurer** [comment]
I volunteer to implement a POC of this :)

**Comment by hmaurer** [comment]
This is clearly getting further than the first implementation of this issue needs to be, but if we take this approach I can see one potential further improvement. Let's call the points breaking up arrows "waypoints". The simplest arrow would be one with no waypoints; a straight arrow.

As @gaearon pointed out such arrows might intersect with shapes along the way; not pretty! One solution would be to move the endpoints of the arrow (still attached to the same shape) to try to eliminate the intersection, but you would have very little wiggle room. Instead, you could manually add a waypoint to the arrow (e.g. by clicking somewhere along its path, and dragging). Then you could generate a diagram similar to the one in my last message, with a broken arrow avoiding the circle.

Now, call this a "manual waypoint". A potentially useful extension to this approach (which should stay simple to implement¹) would be to add "automatic waypoints. A simple rule would be: if an arrow intersects with a shape, add a waypoint along a perpendicular to the line of intersection, like so:

![excalidraw (2)](https://user-images.githubusercontent.com/541045/71785599-640a9a00-2ff9-11ea-9178-06daf7935780.png)

(excalidraw needs lines)

I think this approach could be extended to work around further intersections. It's quite naive (you could implement a full-blown constraint solver but...) but it might give a decent enough result. As I said that's definitely a v2 or v3 of the issue described at the beginning of this ticket though!

¹famous last words

**Comment by vjeux** [comment]
Your last suggestion is basically implementing a layout algorithm. While this is a fun exercise (I may have written a few myself...), I would recommend first implementing the ability to manually move points. It's likely going to cover most use cases. Then if you feel like having fun, you can implement it!

**Comment by hmaurer** [comment]
I am sick at the moment so if someone wants to pick this one up go ahead! :-)

**Comment by gmertk** [comment]
I'll give this a try! 

Very early prototype: 
![Jan-25-2020 17-10-03](https://user-images.githubusercontent.com/907281/73124663-ab030400-3f95-11ea-970e-c449d63dfe3d.gif)

**Comment by dwelle** [comment]
I'm wondering whether we shouldn't give @GasimGasimzada some slack with the arrow changes lately :)

**Comment by gmertk** [comment]
Sure! For now I'm just playing around with the idea. I'd probably wait for his PR to be merged before sending anything serious.

**Comment by vjeux** [comment]
Before you go much further with this, I don’t think that we should only be attach to 4 points for a rectangle. Instead, we should be able to attach anywhere on the edge of the shape. This way if you have multiple arrows pointing to a box, they don’t have to all awkwardly be attached to the center but can take the full width.

In addition, I feel like we should make this “snap” position have some padding, so that the arrow is not drawn directly on the edge which is not very “natural” in real life.

**Comment by j-f1** [comment]
Here’s how Keynote does this: https://github.com/excalidraw/excalidraw/issues/157

**Comment by gmertk** [comment]
Leaving this up for grabs if someone else wants to give it a try.

**Comment by tomayac** [comment]
I know this feature as [glue points](https://wiki.openoffice.org/wiki/Documentation/OOoAuthors_User_Manual/Draw_Guide/Gluepoints_and_connectors) that appear on nodes of shapes or in the middle of edges, but that can also be placed freely.

**Comment by lipis** [comment]
I updated the title :)

**Comment by sompylasar** [comment]
Not sure how, but it would be nice to support the edge case where two rectangles are connected with two or more parallel lines or arrows:

![excalidraw-2020454819](https://user-images.githubusercontent.com/498274/78473171-539ab600-76f3-11ea-8094-4ebda051b5b2.png)

**Comment by lipis** [comment]
@sompylasar As @vjeux pointed already that's the desired solution..

> One thing I don’t like with most implementations is that you only have 3 control points (left, middle, right) and if you have more than a few connectors, then they all are clamped together. I’d love to spread them out all over the surface we’re trying to connect to. Should be easy to implement and greatly improve the rendering.

Also check how Keynote is doing it in one of the comments above.

**Comment by sompylasar** [comment]
@lipis Please forgive me if I'm missing something, but every example I've seen so far does not show 2:2 relationship, they either show one 1:1 or multiple 1:1 cases.

I think my assumption was that the line goes from one center of the figure to another center of the figure, so there can't be two straight lines between two points. They either need to be bent into a curve (and as we know curve editing after creation hasn't been implemented yet), like in Keynote example, or one of them needs to be attached to a third transparent shape with a shifted center.

But now I'm reading these comments once again and there is this you mentioned

> I’d love to spread them out all over the surface we’re trying to connect to. Should be easy to implement and greatly improve the rendering.

And this:

> Instead, we should be able to attach anywhere on the edge of the shape. This way if you have multiple arrows pointing to a box, they don’t have to all awkwardly be attached to the center but can take the full width.

Which should indeed solve my use case.

**Comment by mapbakery** [comment]
I'm hope when this gets implemented that there will be a setting to enable/disable glue points. Because I don't like glue points/automatic layout things. One of the great features of excalidraw is that it is a whiteboard/sketch tool. And it helps the (at least mine) thinking process that I need to erase/rearrange stuff manually (or start all over again sometimes).

**Comment by zyhou** [comment]
Don't know if it can help, [leader-line](https://github.com/anseki/leader-line) seems to solve your issue (demo: https://anseki.github.io/leader-line/)

**Comment by vjeux** [comment]
@mapbakery I'm also very unhappy about the existing status quo regarding this topic. I believe that there's a way to make this feel good. If not, we likely won't ship it.

The best system I found so far is draw.io which gets a lot of things right. It's worth exploring it if you haven't already. The biggest drawback I see is that it assumes that everything is laid out on a grid, whereas excalidraw tends to be more open ended.

**Comment by dwelle** [comment]
Absolutely this would be optional, or would require some user action to connect. We don't want to enable this layouting by default, ala Figma and others, as it'd drastically alter how Excalidraw feels and behaves.

**Comment by mapbakery** [comment]
@vjeux I worked with draw.io but it has issues with glue points when you start moving elements. 

@dwelle Great, thanks for the reassurance :-)

**Comment by xixixao** [comment]
I thought about this for a bit, here's a UX proposal (for simplicity I will go with keyboard and pointer but UI and touch alternatives should be feasible):

1. Select new Arrow or Line shape
2. Hold (macOS) option key *(keep holding until release step)*
3. Hover over any existing shape - the shape will get highlighted similarly to selection mode and the cursor will change
4. Tap to start the line/arrow
5. *For illustration:* another tap outside any shape adds a flex point
6. Tap over another shape places the end of the arrow/line
7. Release option and hit Enter

The arrow/line is now bound to the two shapes.

**Moving/rotating/scaling the bound shapes does the following:**

1. The "binding" point of the arrow/line stays at the same relative "anchor" position to the moved/rotated/scaled shape
2. Any "flex" point is adjusted such that it stays in the same relative position between its neighboring points as before. This should look the same as if the two segments where a shape which was rotated and fixed-ratio scaled such that the unmoved point stays in the same position and the "bound" point moves to its new position *(this likely has a mathematical name I can't remember)*. *(btw this would be nice to have for lines/arrows instead of/in addition to the current default scaling behavior)*

**Editing the line/arrow:**

1. "Bound" points can be moved and they will stay bound to their shapes but their "anchor" position changes
2. When a "bound" point is selected the shape it is bound to is highlighted the same as during creation
3. An "unbound" point can be bound to a shape by holding the option key while moving it
4. A "bound" point can be unbound by holding the option key while moving it (anywhere, including over a shape)

**Copying/Moving the line/arrow:**

1. Copying/moving a line/arrow with bound points keeps those points bound to their shapes but changes their "anchor" position

**Comment by vjeux** [comment]
This sounds very reasonable. I’m curious, wouldn’t it be better to make the “bound” mode be the default? And then if you don’t want it then press a modifier.

I can’t think of many use cases where I would want to draw an arrow next to a shape but not attach it to that shape.

**Comment by xixixao** [comment]
@vjeux playing devil's advocate: I guess some people can share @mapbakery's sentiment: there is some simplicity and straightforward-ness in arrows and (especially) lines behaving like any other shape. And you probably want lines to both not work like that by default AND optionally work like it, and you want lines to work the same as arrows.
But if it felt better...

**Comment by dwelle** [comment]
One problem with the key: `Alt/Option` is used for duplicating during drag, and when tapping on an element you often tend to move a mouse a tiny fraction which will trigger the duplicating.

**Comment by xixixao** [comment]
Actually that wouldn't be a problem for creation (since you're in insertion mode and this rule doesn't apply), but I checked and it also inserts a new point while editing a line, and so that would not work for the editing mode well. Perhaps a different modifier key (command on mac?) could be used?

**Comment by xixixao** [comment]
So I've started to play with a prototype, and also rereading the conversation I think this is an interesting problem, a trade-off between three axes:

1. Simplicity - both how simple the initial user interaction is and how simple the implementation is
2. Control - how precisely can the user define the binding
3. Robustness - a very robust solution doesn't require the user to adjust things as they move things around, while a not a very robust solution requires the user to constantly make manually updates

Here's some versions (which could be combined, and nothing is set in stone yet), that demonstrate this tradeoff:

**Version A (Centroid): Simple, ~Controlled~, Robust**

This is the "attach to shape center" version. This doesn't allow @sompylasar was asking for, because two straight arrows between the same two shapes would overlap.

<img width="800" alt="Screen Shot 2020-07-08 at 10 04 11" src="https://user-images.githubusercontent.com/1473433/86948700-61047980-c102-11ea-9397-fa6da52c3533.png">

**Version B (Glue points): Simple, Controlled, ~Robust~**

This allows attaching to any point on the shape or boundary. It gives the user great control, but it might require adjusting when the bound element is moved, which kinda defies the point.

<img width="737" alt="Screen Shot 2020-07-08 at 10 05 33" src="https://user-images.githubusercontent.com/1473433/86948817-91e4ae80-c102-11ea-8b70-b164300001f5.png">

**Version C (Focus points): ~Simple~, Controlled, Robust**

This version separates the "glue/focus point" from the end of a line. It's basically a combination of the two previous approaches: The user decides where to point (control), but the position of the line end is smartly adjusted like in Version A. There might be several ways to achieve this, and they can tradeoff between simplicity and control, and also they will impact how "nice" the initial interaction feels.

<img width="739" alt="Screen Shot 2020-07-08 at 10 08 08" src="https://user-images.githubusercontent.com/1473433/86949040-ee47ce00-c102-11ea-908f-dcc002e01743.png">

**Comment by vjeux** [comment]
I really like the framing you have with the control point. Here's an idea for a specific algorithm to implement it:

- If you have an arrow that points to a shape
- Find the two points that the arrow interests with the shape
- **Control point**: Center of those two points
- **Gap**: Distance between the end of the arrow and the first intersection

If you move the shape around
- Change the orientation of the arrow to aim at the **control point**
- Find the intersection with the shapes
- Make the distance between the end of the arrow and the first intersection to be the **gap**.

![image](https://user-images.githubusercontent.com/197597/86953585-7cbf4e00-c109-11ea-8f73-c9842530323b.png)

**Comment by vjeux** [comment]
Okay, I spent two hours implementing a prototype for the algorithm above and it looks really good.

https://codesandbox.io/s/quirky-black-3n0q4?file=/src/App.js

![ezgif-2-9799ab1e0e03](https://user-images.githubusercontent.com/197597/86997280-8fb23c80-c162-11ea-9e1d-901c4f0a4656.gif)

Note: I cheated, I was lazy and didn't want to compute the correct position based on the gap, so I added an opaque border the size of the gap around the rectangle. It's not perfect but is an okay approximation. We will want to use a correct implementation for production :)

I'd love for someone to implement it for real :)

**Comment by jayvelasco1990** [comment]
> Okay, I spent two hours implementing a prototype for the algorithm above and it looks really good.
> 
> https://codesandbox.io/s/quirky-black-3n0q4?file=/src/App.js
> 
> ![ezgif-2-9799ab1e0e03](https://user-images.githubusercontent.com/197597/86997280-8fb23c80-c162-11ea-9e1d-901c4f0a4656.gif)
> 
> Note: I cheated, I was lazy and didn't want to compute the correct position based on the gap, so I added an opaque border the size of the gap around the rectangle. It's not perfect but is an okay approximation. We will want to use a correct implementation for production :)
> 
> I'd love for someone to implement it for real :)

I don't think it needs to be perfect. What's shown in your example would fulfill what I would use this tool for. In my circumstance the cosmetic piece is not that important, will just need ability to form relationship between two objects.

**Comment by dwelle** [comment]
fixed via https://github.com/excalidraw/excalidraw/pull/1899 (lines purposefully left out for now --- we'll create a new umbrella issue for follow up improvements, later).

**Comment by lxwbr** [comment]
Can this be turned off? I am having hard time realigning everything, as the algorithm shifts all the lines if I e.g. make a rectangle bigger.

**Comment by dwelle** [comment]
See https://github.com/excalidraw/excalidraw/issues/157 — we can't work out what's the best way to go forward making this opt-in. Ideas welcome.

**Comment by AndreKR** [comment]
Is the binding area the only currently available way or is there a way to switch to glue points?

The binding area is nice if your shapes are all different but if you have a number of similar aligned shapes, it's very hard to also align their arrows.

![image](https://github.com/excalidraw/excalidraw/assets/1188538/10ee3b8d-a76d-451b-be37-e8444d5eb780)

**Comment by dwelle** [comment]
@AndreKR no, but we will be improving this (not yet settled how). But in case, please open a new issue if you don't find one similar already.
